File: packages/core_dart/lib/src/interaction.dart
/// `interaction.dart` — KRO-74. Espejo 1:1 de `interaction.ts`.
///
/// Decisor puro de interactividad: dada una [ViewComposition] y un item de
/// datos, devuelve QUÉ acción ejecutar al tap **sin ejecutarla**. El cliente
/// traduce la [TapResolution] a sus primitivas nativas:
///   - Flutter: Navigator.push() / showModalBottomSheet() / ExpansionTile / launchUrl()
///
/// Pure: sin side effects, sin date/random, sin acceso a UI. Determinístico.
library;

import 'auto_detail.dart';
import 'composition.dart';
import 'field_def.dart';
import 'recipes.dart';

// ─────────────────────────────────────────────────────────────────────────────
// Tipos públicos
// ─────────────────────────────────────────────────────────────────────────────

/// Resultado del resolver de tap: qué debe hacer el cliente al tocar un item.
///
/// Unión sellada (espejo del `type TapResolution` de TS). Variantes:
///   - [TapNone]     — la card no es interactiva, nada pasa.
///   - [TapNavigate] — abrir pantalla completa con `detailComposition`.
///   - [TapModal]    — abrir overlay/bottom-sheet con `detailComposition`.
///   - [TapExpand]   — desplegar accordion inline (receta + slots resueltos).
///   - [TapExternal] — abrir URL externa (puede estar vacía si el field no tiene valor).
sealed class TapResolution {
  const TapResolution();

  /// Discriminante textual, paridad con `{ kind: ... }` del TS.
  String get kind;
}

class TapNone extends TapResolution {
  const TapNone();
  @override
  String get kind => 'none';
}

class TapNavigate extends TapResolution {
  final ViewComposition detailComposition;
  const TapNavigate(this.detailComposition);
  @override
  String get kind => 'navigate';
}

class TapModal extends TapResolution {
  final ViewComposition detailComposition;
  const TapModal(this.detailComposition);
  @override
  String get kind => 'modal';
}

class TapExpand extends TapResolution {
  final String expandRecipe;
  final Map<String, SlotComposition> expandSlots;
  const TapExpand({required this.expandRecipe, required this.expandSlots});
  @override
  String get kind => 'expand';
}

class TapExternal extends TapResolution {
  final String url;
  const TapExternal(this.url);
  @override
  String get kind => 'external';
}

/// Opciones de [resolveTapAction]. Espejo de `ResolveTapOptions` (TS).
class ResolveTapOptions {
  /// Definiciones de fields de la sección. Si se proveen, se usan para construir
  /// la `detailComposition` automática cuando el publisher no declaró
  /// `targetComposition` ni `targetRecipe`. Sin ellos, los slots del detalle
  /// quedan vacíos (el renderer caerá a su fallback visual).
  final List<FieldDefLike> fieldDefs;
  const ResolveTapOptions({this.fieldDefs = const []});
}

// ─────────────────────────────────────────────────────────────────────────────
// API pública
// ─────────────────────────────────────────────────────────────────────────────

/// Dada una [composition] y el [item] de datos tocado, devuelve la
/// [TapResolution] — qué debe hacer el cliente.
///
/// El [item] solo se usa para `external_link` (extraer la URL del `linkField`).
TapResolution resolveTapAction(
  ViewComposition composition,
  Map<String, dynamic> item, [
  ResolveTapOptions options = const ResolveTapOptions(),
]) {
  final fieldDefs = options.fieldDefs;

  switch (composition.action) {
    case 'navigate_to_detail':
      return TapNavigate(resolveDetailComposition(composition, fieldDefs));

    case 'modal':
      return TapModal(resolveDetailComposition(composition, fieldDefs));

    case 'expand_inline':
      final expand = composition.expand;
      if (expand == null) return const TapNone();
      return TapExpand(expandRecipe: expand.recipe, expandSlots: expand.slots);

    case 'external_link':
      final raw =
          composition.linkField != null ? item[composition.linkField] : null;
      final url = raw is String ? raw.trim() : '';
      return TapExternal(url);

    case 'none':
    default:
      return const TapNone();
  }
}

/// Resuelve la mejor [ViewComposition] para mostrar como pantalla de detalle
/// (destino de `navigate_to_detail` o `modal`).
///
/// Prioridad:
///   1. `targetComposition` (cadena KRO-94): receta + slots del primer salto; si
///      sus slots están vacíos, se rellenan con `buildAutoDetailComposition`.
///      Los saltos más profundos se preservan para encadenar navegación.
///   2. `targetRecipe` (legacy, solo si es de `kind:'detail'`): esa receta + auto-slots.
///   3. Fallback: `buildAutoDetailComposition(fieldDefs)` (`hero_protagonico`).
ViewComposition resolveDetailComposition(
  ViewComposition composition, [
  List<FieldDefLike> fieldDefs = const [],
]) {
  // 1. Cadena multi-salto (KRO-94).
  final hop = composition.targetComposition;
  if (hop != null) {
    final autoSlots = buildAutoDetailComposition(fieldDefs).slots;
    final slots = (hop.slots != null && hop.slots!.isNotEmpty)
        ? hop.slots!
        : autoSlots;
    return ViewComposition(
      recipe: hop.recipe,
      action: hop.action,
      slots: slots,
      targetComposition: hop.targetComposition,
      expand: hop.expand,
      linkField: hop.linkField,
    );
  }

  // 2. Legacy targetRecipe.
  final targetRecipe = composition.targetRecipe;
  if (targetRecipe != null) {
    final manifest = getRecipeManifest(targetRecipe);
    if (manifest != null && manifest.kind == 'detail') {
      final auto = buildAutoDetailComposition(fieldDefs);
      return ViewComposition(
        recipe: targetRecipe,
        action: auto.action,
        slots: auto.slots,
      );
    }
  }

  // 3. Fallback auto.
  return buildAutoDetailComposition(fieldDefs);
}

/// Auto-elige la mejor receta de DETALLE para los fields de una sección.
/// galería → `momento`; imagen prominente → `hero_protagonico`; resto → `editorial`.
String resolveTargetRecipe(List<FieldDefLike> fieldDefs) {
  final hasGallery = fieldDefs.any((f) =>
      const ['gallery', 'slideshow', 'card_multiview'].contains(f.behavior ?? '') ||
      f.type == 'array<image>');
  if (hasGallery) return 'momento';

  final hasMainImage = fieldDefs.any((f) =>
      const ['avatar', 'banner', 'cover', 'thumbnail'].contains(f.behavior ?? '') ||
      f.type == 'image');
  if (hasMainImage) return 'hero_protagonico';

  return 'editorial';
}

/// Auto-elige la mejor receta de despliegue (expand): con fields enlazables
/// (url/email/phone) → `accordion_with_actions`; solo texto → `accordion_simple`.
String resolveExpandRecipe(List<FieldDefLike> fieldDefs) {
  final hasActionable = fieldDefs
      .any((f) => const ['url', 'email', 'phone'].contains(f.behavior ?? ''));
  return hasActionable ? 'accordion_with_actions' : 'accordion_simple';
}

/// ¿El tap es navegable o no? (cursor "pointer" solo si hay algo que tocar).
bool isTappable(ViewComposition composition) => composition.action != 'none';

/// ¿La acción del primer tap abre una pantalla nueva (vs inline)?
/// `navigate_to_detail` y `modal` abren pantalla; `expand_inline` y
/// `external_link` no salen de la lista.
bool opensNewScreen(ViewComposition composition) =>
    composition.action == 'navigate_to_detail' || composition.action == 'modal';

File: packages/core/src/interaction.ts
/**
 * `interaction.ts` — KRO-74 (B++).
 *
 * Decisor puro de interactividad: dado una `ViewComposition` y un item de datos,
 * devuelve QUÉ acción ejecutar al tap sin ejecutarla. Los clientes traducen la
 * resolución a sus primitivas nativas:
 *   - Studio:  animación in-frame / modal slide-up / accordion / toast URL
 *   - Flutter: Navigator.push() / showModalBottomSheet() / ExpansionTile / launchUrl()
 *
 * Esto cierra el ciclo del SDK polyglot:
 *   1. Modelo  (qué entidades existen)       → `@kromia/core` ✅
 *   2. Datos sintéticos (fake data preview)  → `synthSectionItems` ✅  (KRO-72)
 *   3. Presentación  (format / accent / auto)→ `formatScalar` / `extractAccentSettings` ✅ (KRO-73)
 *   4. Interactividad (qué pasa al tocar)    → `resolveTapAction` ✅  ← esta tarea
 *
 * Pure: sin side effects, sin date/random, sin acceso a UI. Determinístico.
 */

import type { ViewComposition, SlotComposition, RecipeId } from './types';
import type { FieldDefLike } from './types';
import { resolveTargetChain } from './target-chain';
import { buildAutoDetailComposition } from './auto-detail';
import { getRecipeManifest } from './registries/recipes';

// ─────────────────────────────────────────────────────────────────────────────
// Tipos públicos
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Resultado del resolver de tap: qué debe hacer el cliente al tocar un item.
 *
 * `none`     — la card no es interactiva, nada pasa.
 * `navigate` — abrir pantalla completa con `detailComposition`.
 * `modal`    — abrir overlay/bottom-sheet con `detailComposition`.
 * `expand`   — desplegar accordion inline (receta + slots ya resueltos).
 * `external` — abrir URL externa (puede estar vacía si el field no tiene valor).
 */
export type TapResolution =
  | { kind: 'none' }
  | { kind: 'navigate'; detailComposition: ViewComposition }
  | { kind: 'modal';    detailComposition: ViewComposition }
  | { kind: 'expand';   expandRecipe: RecipeId; expandSlots: Record<string, SlotComposition> }
  | { kind: 'external'; url: string };

export interface ResolveTapOptions {
  /**
   * Definiciones de fields de la sección. Si se proveen, se usan para
   * construir la `detailComposition` automática cuando el publisher no
   * declaró `targetComposition` ni `targetRecipe`. Sin ellos, los slots
   * del detalle quedan vacíos (el renderer caerá a su fallback visual).
   */
  fieldDefs?: FieldDefLike[];
}

// ─────────────────────────────────────────────────────────────────────────────
// API pública
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Dado una `ViewComposition` y el item de datos tocado, devuelve la
 * `TapResolution` — qué debe hacer el cliente.
 *
 * Compatibilidad:
 *  - `targetComposition` (KRO-94, cadena multi-salto) tiene prioridad.
 *  - `targetRecipe` (legacy single-hop) se honra si no hay cadena.
 *  - Si no hay nada → `buildAutoDetailComposition(fieldDefs)` como fallback.
 *
 * El `item` solo se usa para `external_link` (extraer la URL del `linkField`).
 * Para los otros tipos de acción el item no afecta a la resolución.
 */
export function resolveTapAction(
  composition: ViewComposition,
  item: Record<string, unknown>,
  options: ResolveTapOptions = {},
): TapResolution {
  const { fieldDefs = [] } = options;

  switch (composition.action) {
    case 'navigate_to_detail':
      return {
        kind: 'navigate',
        detailComposition: resolveDetailComposition(composition, fieldDefs),
      };

    case 'modal':
      return {
        kind: 'modal',
        detailComposition: resolveDetailComposition(composition, fieldDefs),
      };

    case 'expand_inline': {
      const expand = composition.expand;
      if (!expand) return { kind: 'none' };
      return {
        kind: 'expand',
        expandRecipe: expand.recipe,
        expandSlots: expand.slots,
      };
    }

    case 'external_link': {
      const raw = composition.linkField ? item[composition.linkField] : undefined;
      const url = typeof raw === 'string' ? raw.trim() : '';
      return { kind: 'external', url };
    }

    case 'none':
    default:
      return { kind: 'none' };
  }
}

/**
 * Resuelve la mejor `ViewComposition` para mostrar como pantalla de detalle
 * (destino de `navigate_to_detail` o `modal`).
 *
 * Prioridad:
 *   1. `targetComposition` (cadena KRO-94): usa la receta + slots del primer
 *      salto de la cadena; si sus slots están vacíos, los rellena con
 *      `buildAutoDetailComposition(fieldDefs)`. Los saltos más profundos
 *      (`targetComposition.targetComposition`) se preservan tal cual para que
 *      el renderer pueda encadenar navegación.
 *   2. `targetRecipe` (legacy): receta declarada + auto-slots de fieldDefs.
 *   3. Fallback: `buildAutoDetailComposition(fieldDefs)` con `hero_protagonico`.
 */
export function resolveDetailComposition(
  composition: ViewComposition,
  fieldDefs: FieldDefLike[] = [],
): ViewComposition {
  // 1. Cadena multi-salto (KRO-94).
  if (composition.targetComposition) {
    const hop = composition.targetComposition;
    // KRO-317 (2ª parte) — los slots sintetizados son PARA la receta del salto.
    // Sin pasarla, `buildAutoDetailComposition` los construía para la receta por
    // defecto (`hero_protagonico`) y se los encajábamos a otra distinta: los
    // slots que esa receta no tiene se quedan huérfanos y los que sí tiene se
    // quedan vacíos. Es el mismo fallo que se arregló abajo, en la rama de
    // `targetRecipe`, y que en el detalle de sección se veía como una portada en
    // blanco. Sobrevivía aquí porque esta rama es la multi-salto (KRO-94) y se
    // arregló solo la otra.
    const autoSlots = buildAutoDetailComposition(fieldDefs, hop.recipe).slots;
    const slots = hop.slots && Object.keys(hop.slots).length > 0
      ? hop.slots
      : autoSlots;
    return {
      recipe:  hop.recipe,
      action:  hop.action,
      slots,
      // KRO-133 — el DETALLE puede tener su propio diseño por BLOQUES
      // (`layout`) + slots deshabilitados/custom (`slotOverrides`) + posición de
      // acento. El tipo `TargetComposition` los declara y Studio los persiste,
      // pero antes se PERDÍAN aquí → el cliente (Flutter) caía siempre al preset
      // en vez de pintar el lienzo guardado. Se propagan tal cual.
      ...(hop.layout         ? { layout: hop.layout }                 : {}),
      ...(hop.slotOverrides  ? { slotOverrides: hop.slotOverrides }   : {}),
      ...(hop.accentPosition ? { accentPosition: hop.accentPosition } : {}),
      ...(hop.targetComposition ? { targetComposition: hop.targetComposition } : {}),
      ...(hop.expand            ? { expand: hop.expand }                         : {}),
      ...(hop.linkField         ? { linkField: hop.linkField }                   : {}),
    };
  }

  // 2. Legacy targetRecipe.
  if (composition.targetRecipe) {
    const manifest = getRecipeManifest(composition.targetRecipe);
    if (manifest && manifest.kind === 'detail') {
      // KRO-317 — la receta destino se le PASA al constructor.
      //
      // Antes se llamaba sin ella, así que la heurística legacy producía slots
      // de `hero_protagonico` (`banner`, `avatar`, `subtitle`…) y acto seguido se
      // les pegaba encima la etiqueta de OTRA receta. Editorial iba entonces a
      // buscar su `cover` a una composición que no lo tenía → **portada vacía**,
      // y lo mismo con el `slideshow` de momento.
      //
      // Studio arregló exactamente esto en su lado (KRO-131) y dejó el motivo
      // escrito en `SectionAppPreview.tsx`; esta copia del SDK se quedó con la
      // versión rota. Mismo fallo corregido en un sitio y olvidado en el de al
      // lado — que es como se ven distintos el AppPreview y la app pintando la
      // MISMA sección.
      return {
        ...buildAutoDetailComposition(fieldDefs, composition.targetRecipe),
        recipe: composition.targetRecipe,
      };
    }
  }

  // 3. Fallback auto.
  return buildAutoDetailComposition(fieldDefs);
}

/**
 * KRO-74 — Auto-elige la mejor receta de detalle para los fields de una sección.
 *
 * `hero_protagonico` cuando hay imagen prominente (avatar/banner/cover/thumbnail
 * o `type=image`). `editorial` cuando la sección es mayormente texto/fecha/stats
 * sin imagen dominante. `momento` cuando hay galería de imágenes.
 *
 * Pure: `(fieldDefs) => RecipeId`. Determinístico.
 */
export function resolveTargetRecipe(fieldDefs: FieldDefLike[]): RecipeId {
  const hasGallery = fieldDefs.some(f =>
    ['gallery', 'slideshow', 'card_multiview'].includes(f.behavior ?? '') ||
    f.type === 'array<image>',
  );
  if (hasGallery) return 'momento';

  const hasMainImage = fieldDefs.some(f =>
    ['avatar', 'banner', 'cover', 'thumbnail'].includes(f.behavior ?? '') ||
    f.type === 'image',
  );
  if (hasMainImage) return 'hero_protagonico';

  return 'editorial';
}

/**
 * KRO-74 — Auto-elige la mejor receta de despliegue (expand) para los fields
 * de una sección.
 *
 * `accordion_with_actions` cuando hay fields de enlace (url/email/phone) que
 * el accordion puede convertir en botones de acción. `accordion_simple` si
 * solo hay texto.
 *
 * Pure: `(fieldDefs) => RecipeId`. Determinístico.
 */
export function resolveExpandRecipe(fieldDefs: FieldDefLike[]): RecipeId {
  const hasActionable = fieldDefs.some(f =>
    ['url', 'email', 'phone'].includes(f.behavior ?? ''),
  );
  return hasActionable ? 'accordion_with_actions' : 'accordion_simple';
}

/**
 * Conveniencia: dado una `ViewComposition`, ¿el tap es navegable o no?
 * Útil para que el renderer de lista muestre el cursor "pointer" solo
 * cuando hay algo que tocar.
 */
export function isTappable(composition: ViewComposition): boolean {
  return composition.action !== 'none';
}

/**
 * Conveniencia: ¿la acción del primer tap abre una pantalla nueva (vs inline)?
 * `navigate_to_detail` y `modal` abren pantalla; `expand_inline` y
 * `external_link` no salen de la lista.
 */
export function opensNewScreen(composition: ViewComposition): boolean {
  return (
    composition.action === 'navigate_to_detail' ||
    composition.action === 'modal'
  );
}
